streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java

```diff
@@ -0,0 +1,459 @@
+        describe(builder, suspended.values(), indent, "Suspended:");
+        describe(builder, restoring.values(), indent, "Restoring:");
+        describe(builder, created.values(), indent, "New:");
+        return builder.toString();
+    }
+
+    private void describe(final StringBuilder builder,
+                          final Collection<Task> tasks,
+                          final String indent,
+                          final String name) {
+        builder.append(indent).append(name);
+        for (final Task t : tasks) {
+            builder.append(indent).append(t.toString(indent + "\t\t"));
+        }
+        builder.append("\n");
+    }
+
+    private List<Task> allInitializedTasks() {
+        final List<Task> tasks = new ArrayList<>();
+        tasks.addAll(running.values());
+        tasks.addAll(suspended.values());
+        tasks.addAll(restoring.values());
+        return tasks;
+    }
+
+    Collection<Task> restoringTasks() {
+        return Collections.unmodifiableCollection(restoring.values());
+    }
+
+    Set<TaskId> allAssignedTaskIds() {
+        final Set<TaskId> taskIds = new HashSet<>();
+        taskIds.addAll(running.keySet());
+        taskIds.addAll(suspended.keySet());
+        taskIds.addAll(restoring.keySet());
+        taskIds.addAll(created.keySet());
+        return taskIds;
+    }
+
+    void clear() {
+        runningByPartition.clear();
+        running.clear();
+        created.clear();
+        suspended.clear();
+        restoredPartitions.clear();
+    }
+
+    Set<TaskId> previousTaskIds() {
+        return previousActiveTasks;
+    }
+
+    void commit() {
+        performTimedAction(commitAction);
+    }
+
+    void maybeCommit() {
+        performTimedAction(maybeCommitAction);
+    }
+
+    private void performTimedAction(final TaskAction action) {
+        timerStartedMs = time.milliseconds();
+        final RuntimeException exception = applyToRunningTasks(action, false);
+        if (exception != null) {
+            throw exception;
+        }
+    }
+
+    int process() {
+        timerStartedMs = time.milliseconds();
+        int processed = 0;
+        timerStartedMs = time.milliseconds();
+        for (final Task task : running()) {
+            if (task.process()) {
+                processSensor.record(computeLatency(), timerStartedMs);
+                processed++;
+                try {
+                    if (task.maybePunctuateStreamTime()) {
+                        punctuateSensor.record(computeLatency(), timerStartedMs);
+                    }
+                } catch (final KafkaException e) {
+                    log.error("{} Failed to punctuate active task {} due to the following error:", logPrefix, task.id(), e);
+                    throw e;
+                }
+            }
+        }
+        return processed;
+    }
+
+    RuntimeException applyToRunningTasks(final TaskAction action, final boolean throwException) {
```

Comment:
Are you saying add `punctuate` and `print` methods to `AssignedTasks`?  Sounds reasonable